qiskit/dagcircuit/dagcircuit.py

```diff
@@ -511,11 +511,16 @@ def _map_condition(self, wire_map, condition):
         if condition is None:
             new_condition = None
         else:
-            # Map the register name, using fact that registers must not be
-            # fragmented by the wire_map (this must have been checked
-            # elsewhere)
-            bit0 = condition[0][0]
-            new_condition = (wire_map.get(bit0, bit0).register, condition[1])
+            # if there is a condition, map the condition bits to the
+            # composed cregs based on the wire_map
+            cond_val = condition[1]
+            new_cond_val = 0
+            for bit in wire_map:
+                if isinstance(bit, Clbit):
+                    new_creg = wire_map[bit].register
+                    if 2**(bit.index) & cond_val:
+                        new_cond_val += 2**(wire_map[bit].index)
+            new_condition = (new_creg, new_cond_val)
```

Comment:
`DAGCircuit.compose` won't always populate `wire_map` with all of the `Clbit`s used in the `condition`, in which case we'll end up raising a name unbound error (example below).

I think the right fix is to update `compose` to always fill the needed edges in `wire_map` from the identity wire map (though we could alternately update here to handle missing edges).

```
>>> qc = qk.QuantumCircuit(2,1)
>>> qc.h(0)
>>> qc.x(1).c_if(qc.cregs[0], 1)
>>> qc_2 = qk.QuantumCircuit(2,1)
>>> qc_2.compose(qc, qubits=[1,0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 518, in compose
    dag_self.compose(dag_other, qubits=qubits, clbits=clbits, front=front)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/dagcircuit/dagcircuit.py", line 628, in compose
    condition = dag._map_condition(edge_map, nd.condition)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/dagcircuit/dagcircuit.py", line 523, in _map_condition
    new_condition = (new_creg, new_cond_val)
UnboundLocalError: local variable 'new_creg' referenced before assignment
```

Reply:
I figured there might be some error handling required here. I'll look at `compose` and see what it takes to put make sure the Clbit's are there.

Reply:
@kdk. Just to confirm - the reason for the error is that there is only a qubits= parameter to compose and therefore the clbit for the condition is not carried into the wire_map. The solution would be to always check for condition and if there, add the necessary clbits to the wire_map.
If that's done, should we still trap an error before line 523 above if new_creg is not valid? Could set new_creg to None at the top and if None at 523, error or set new_condition to None?

Reply:
That's right, and the proposed solution sounds good. We should still check that new_creg is valid (and raise if not) since there are other callers of `_map_condition`.

Reply:
Ok, using the `identity_clbit_map`, I can add what `clbit `is needed in your example above and works fine. Where it gets a bit gnarly is what if in your example from #4615, the user enters `clbits=[1]`. What should 0 & 2 map to? Just go in order?

Suppose 10 `clbits` and user enters `clbits=[1, 4, 8, 6]`. Could get messy in the general case and should we be making assumptions for the user or should we just raise an error if condition does not have a `clbit `declared?

Reply:
If the user provides a `qubits`/`clbits` argument, we should check that its length matches the number of qubits/clbits in `other`, and raise if not. (This had been handled by `_check_edgemap_registers`, but wasn't carried over when `compose` was written.) When one of them is `None`, we can substitute in the full `identity_{qubit,clbit}_map`.

Reply:
Great. That sounds like a straightforward solution. I'll take care of that and push the change.